File: SexyBiscuit.Engine/UI/UiInput.cs
using Microsoft.Xna.Framework;

namespace SexyBiscuit.Engine.UI;

/// <summary>
/// One frame of input, in the form the UI needs it. The mirror of the object
/// <c>html5/src/ui/UiInput.js</c> takes.
/// </summary>
/// <remarks>
/// A plain struct rather than a reference to the <c>InputManager</c> on purpose. The router
/// is then a pure function of this and the tree, so a fixture can drive it with no window,
/// no device and no gamepad — which is the only way the two engines' interaction behaviour
/// can be pinned by the same file rather than by two hand-written test suites that drift.
/// </remarks>
public struct UiInputFrame
{
    /// <summary>Unscaled seconds since the last frame. Drives the caret and the mode tracker.</summary>
    public float DeltaTime;

    /// <summary>Pointer position in device pixels. Off-screen is fine; nothing will be hit.</summary>
    public Vector2 Pointer;

    /// <summary>Pointer movement since the last frame, in device pixels.</summary>
    public Vector2 PointerDelta;

    /// <summary>Whether the primary button or the finger is held. Held, not pressed.</summary>
    public bool PointerDown;

    /// <summary>Whether the pointer is a finger, which has no hover state to show.</summary>
    public bool PointerIsTouch;

    /// <summary>Wheel notches this frame. Positive scrolls the content up, as every OS does.</summary>
    public float Wheel;

    /// <summary>Navigation stick, each axis -1..1. Feeds the input-mode hysteresis.</summary>
    public Vector2 NavAxis;

    /// <summary>Discrete navigation presses, from a D-pad, the arrow keys or a stick edge.</summary>
    public bool NavUp, NavDown, NavLeft, NavRight;

    /// <summary>Activate the focused node — A, Enter or Space.</summary>
    public bool Confirm;

    /// <summary>Back out — B or Escape. Closes an open list before it does anything else.</summary>
    public bool Cancel;

    /// <summary>Characters typed this frame, for whichever text field has focus.</summary>
    public string Typed;

    /// <summary>Whether backspace was pressed this frame.</summary>
    public bool Backspace;
}

/// <summary>
/// Turns a frame of input into hover, press, click, focus and control behaviour on one
/// <see cref="UiCanvas"/>. The mirror of <c>html5/src/ui/UiInput.js</c>.
/// </summary>
/// <remarks>
/// <para>
/// The seven kinds after <see cref="UiKind.Image"/> have no painter of their own and no
/// behaviour of their own either — a <see cref="UiKind.Slider"/> was a tag that made a node
/// focusable and nothing more. This is where dragging, toggling, opening and scrolling
/// actually happen, so that a document describing a settings screen becomes a settings
/// screen rather than a picture of one.
/// </para>
/// <para>
/// A click is a release inside the node the press started on. Dragging off a button and
/// letting go must not fire it, which is what every other toolkit means by a click.
/// </para>
/// </remarks>
public sealed class UiInput
{
    private readonly UiCanvas _canvas;

    /// <summary>The node the current press started on, which is the only one that can click.</summary>
    private UiNode? _pressed;

    /// <summary>The slider or scroll bar the pointer is dragging, if any.</summary>
    private UiNode? _dragging;

    private bool    _wasDown;
    private Vector2 _lastPointer;

    public UiInput(UiCanvas canvas)
    {
        _canvas = canvas;
        Focus   = new UiFocus(canvas.Root);
    }

    /// <summary>Focus and the input-mode tracker for this canvas.</summary>
    public UiFocus Focus { get; }

    /// <summary>The node under the pointer this frame, or null.</summary>
    public UiNode? Hovered { get; private set; }

    /// <summary>How far one wheel notch scrolls, in canvas units.</summary>
    public float WheelStep { get; set; } = 48f;

    // -------------------------------------------------------------------------
    // The frame
    // -------------------------------------------------------------------------

    /// <summary>
    /// Resolves one frame. Call after the canvas has been laid out, so every
    /// rectangle the hit test reads is this frame's.
    /// </summary>
    public void Update(UiInputFrame frame)
    {
        ClearTransient(_canvas.Root);
        Focus.Modes.Tick(frame.DeltaTime);

        if (!_canvas.Interactive)
        {
            _wasDown     = false;
            _pressed     = null;
            _dragging    = null;
            Hovered      = null;
            return;
        }

        NoteDevices(frame);

        // Before anything reads it. Focus resolved afterwards would mean the first frame
        // a menu is open swallows its own input: a key press with nothing focused yet has
        // nowhere to go, and the player has to press twice.
        Focus.Repair();

        UiNode? hit = Pick(frame.Pointer);
        Hovered = Focus.Modes.ShowHover ? hit : null;
        if (Hovered != null) Hovered.Hovered = true;

        Pointer(frame, hit);
        Wheel(frame, hit);
        Directional(frame);
        Typing(frame);

        SyncFocusFlags(_canvas.Root);

        _wasDown     = frame.PointerDown;
        _lastPointer = frame.Pointer;
    }

    // -------------------------------------------------------------------------
    // Devices
    // -------------------------------------------------------------------------

    private void NoteDevices(UiInputFrame frame)
    {
        if (frame.PointerIsTouch && frame.PointerDown) Focus.Modes.NoteTouch();
        else if (frame.PointerDelta != Vector2.Zero)   Focus.Modes.NoteMouseMotion(frame.PointerDelta);

        if (frame.PointerDown && !_wasDown && !frame.PointerIsTouch) Focus.Modes.NoteMouseButton();
        if (frame.Wheel != 0f) Focus.Modes.NoteMouseButton();

        if (frame.NavAxis != Vector2.Zero)
            Focus.Modes.NoteNavigationAxis(MathF.Max(MathF.Abs(frame.NavAxis.X), MathF.Abs(frame.NavAxis.Y)));

        if (frame.NavUp || frame.NavDown || frame.NavLeft || frame.NavRight || frame.Confirm)
            Focus.Modes.NoteNavigationButton();
    }

    // -------------------------------------------------------------------------
    // Pointer
    // -------------------------------------------------------------------------

    /// <summary>
    /// The node under a device-pixel point, an open dropdown list winning over the tree.
    /// </summary>
    /// <remarks>
    /// A list is painted outside its control's rectangle and outside every clip, so the
    /// ordinary hit test cannot see it. Without this a player can see the options and
    /// clicks straight through them onto whatever is behind.
    /// </remarks>
    private UiNode? Pick(Vector2 devicePoint)
    {
        Vector2 point = _canvas.ScreenToCanvas(devicePoint);

        if (ExpandedDropdown() is { } open && UiPainter.DropdownListRect(open).Contains(point))
            return open;

        return _canvas.HitTest(devicePoint);
    }

    private void Pointer(UiInputFrame frame, UiNode? hit)
    {
        bool pressedNow  = frame.PointerDown && !_wasDown;
        bool releasedNow = !frame.PointerDown && _wasDown;

        if (pressedNow) BeginPress(frame, hit);
        else if (frame.PointerDown) ContinuePress(frame);
        else if (releasedNow) EndPress(frame, hit);

        if (_pressed != null) _pressed.Pressed = frame.PointerDown;
    }

    private void BeginPress(UiInputFrame frame, UiNode? hit)
    {
        Vector2 point = _canvas.ScreenToCanvas(frame.Pointer);

        // A press that lands anywhere but the open list closes it, including on the
        // control itself — which is what makes a second click on a dropdown shut it.
        if (ExpandedDropdown() is { } open && open != hit) open.Expanded = false;

        // A disabled node still blocks the pointer — it is not a hole — but it must not
        // press, click or take focus, so the press lands on nothing at all.
        _pressed  = Usable(hit) ? hit : null;
        _dragging = null;
        if (_pressed is null || hit is null) return;

        if (UiFocus.IsFocusable(hit)) Focus.Focus(hit);

        switch (hit.Kind)
        {
            case UiKind.Slider:
                _dragging = hit;
                SetSliderFromPointer(hit, point);
                break;

            case UiKind.TabStrip:
                SelectTabAt(hit, point);
                break;

            case UiKind.Dropdown when hit.Expanded:
                SelectOptionAt(hit, point);
                break;
        }
    }

    /// <summary>
    /// Whether a press is currently being dragged across a control.
    /// </summary>
    /// <remarks>
    /// Read by <see cref="UiCanvas.ScreenToCanvas"/> in world space, where a pointer ray can
    /// stop meeting the canvas's plane mid-drag if the camera swings. Holding the last good
    /// point is better than reporting a miss, which would read as a jump to the far edge.
    /// </remarks>
    internal bool IsDragging => _dragging != null;

    private void ContinuePress(UiInputFrame frame)
    {
        if (_dragging is { Kind: UiKind.Slider } slider)
            SetSliderFromPointer(slider, _canvas.ScreenToCanvas(frame.Pointer));
    }

    private void EndPress(UiInputFrame frame, UiNode? hit)
    {
        UiNode? pressed = _pressed;
        _pressed  = null;
        _dragging = null;

        if (pressed is null) return;
        pressed.Pressed = false;

        // The release has to land on the same node the press started on.
        if (!ReferenceEquals(pressed, hit)) return;

        pressed.Clicked = true;
        Activate(pressed, ActivationSource.Pointer);
    }

    private void SetSliderFromPointer(UiNode slider, Vector2 point)
    {
        if (slider.Rect.Width <= 0f) return;

        float fraction = Math.Clamp((point.X - slider.Rect.X) / slider.Rect.Width, 0f, 1f);
        SetValue(slider, slider.MinValue + (slider.MaxValue - slider.MinValue) * fraction);
    }

    /// <summary>Whether a node and every ancestor of it is visible and interactive.</summary>
    private static bool Usable(UiNode? node)
    {
        for (UiNode? n = node; n != null; n = n.Parent)
            if (!n.Visible || !n.Interactive) return false;
        return node != null;
    }

    private static void SelectTabAt(UiNode strip, Vector2 point)
    {
        for (int i = 0; i < strip.Options.Count; i++)
            if (UiPainter.TabRect(strip, i).Contains(point)) { strip.SelectedIndex = i; return; }
    }

    private static void SelectOptionAt(UiNode dropdown, Vector2 point)
    {
        RectangleF list = UiPainter.DropdownListRect(dropdown);
        if (dropdown.Rect.Height <= 0f || !list.Contains(point)) return;

        int index = (int)((point.Y - list.Y) / dropdown.Rect.Height);
        if (index < 0 || index >= dropdown.Options.Count) return;

        dropdown.SelectedIndex = index;
        dropdown.Expanded      = false;
        dropdown.Clicked       = true;
    }

    // -------------------------------------------------------------------------
    // Wheel
    // -------------------------------------------------------------------------

    private void Wheel(UiInputFrame frame, UiNode? hit)
    {
        if (frame.Wheel == 0f || hit is null) return;
        if (NearestScrollable(hit) is not { } view) return;

        Vector2 max = ScrollRange(view);
        bool vertical = view.Scroll is ScrollMode.Vertical or ScrollMode.Both && max.Y > 0f;

        Vector2 offset = view.ScrollOffset;
        if (vertical) offset.Y = Math.Clamp(offset.Y - frame.Wheel * WheelStep, 0f, max.Y);
        else          offset.X = Math.Clamp(offset.X - frame.Wheel * WheelStep, 0f, max.X);

        view.ScrollOffset = offset;
    }

    /// <summary>The first ancestor that scrolls and has something to scroll over.</summary>
    private static UiNode? NearestScrollable(UiNode from)
    {
        for (UiNode? n = from; n != null; n = n.Parent)
        {
            if (n.Scroll == ScrollMode.None) continue;
            Vector2 range = ScrollRange(n);
            if (range.X > 0f || range.Y > 0f) return n;
        }
        return null;
    }

    private static Vector2 ScrollRange(UiNode view) => new(
        MathF.Max(0f, view.ContentSize.X - view.ContentRect.Width),
        MathF.Max(0f, view.ContentSize.Y - view.ContentRect.Height));

    // -------------------------------------------------------------------------
    // Directional
    // -------------------------------------------------------------------------

    private void Directional(UiInputFrame frame)
    {
        if (frame.Cancel && ExpandedDropdown() is { } open)
        {
            open.Expanded = false;
            return;
        }

        UiNode? focused = Focus.Focused;

        // An open list swallows up and down: they move through the options, and only
        // once it is closed do they move between controls again.
        if (focused is { Kind: UiKind.Dropdown, Expanded: true })
        {
            if (frame.NavUp)   StepSelection(focused, -1);
            if (frame.NavDown) StepSelection(focused, +1);
            if (frame.Confirm) { focused.Expanded = false; focused.Clicked = true; }
            return;
        }

        // A focused slider takes left and right as its own, because navigating away
        // from a volume slider by pressing right is not what anybody means by it.
        if (focused is { Kind: UiKind.Slider } slider && (frame.NavLeft || frame.NavRight))
        {
            float step = slider.Step > 0f ? slider.Step : (slider.MaxValue - slider.MinValue) / 20f;
            SetValue(slider, slider.Value + (frame.NavRight ? step : -step));
            return;
        }

        // So does a tab strip, which is a row of choices however it is laid out.
        if (focused is { Kind: UiKind.TabStrip } strip && (frame.NavLeft || frame.NavRight))
        {
            StepSelection(strip, frame.NavRight ? +1 : -1);
            return;
        }

        if (frame.NavUp)    Focus.Navigate(NavDirection.Up);
        if (frame.NavDown)  Focus.Navigate(NavDirection.Down);
        if (frame.NavLeft)  Focus.Navigate(NavDirection.Left);
        if (frame.NavRight) Focus.Navigate(NavDirection.Right);

        if (frame.Confirm && Focus.Focused is { } target)
        {
            target.Clicked = true;
            Activate(target, ActivationSource.Directional);
        }
    }

    private static void StepSelection(UiNode node, int by)
    {
        int count = node.Options.Count;
        if (count == 0) return;
        node.SelectedIndex = Math.Clamp(node.SelectedIndex + by, 0, count - 1);
    }

    // -------------------------------------------------------------------------
    // Typing
    // -------------------------------------------------------------------------

    private void Typing(UiInputFrame frame)
    {
        if (Focus.Focused is not { Kind: UiKind.TextField } field) return;

        if (frame.Backspace && field.Text.Length > 0)
            field.Text = field.Text[..^1];

        if (!string.IsNullOrEmpty(frame.Typed))
            field.Text += frame.Typed;
    }

    // -------------------------------------------------------------------------
    // Activation
    // -------------------------------------------------------------------------

    private enum ActivationSource { Pointer, Directional }

    /// <summary>
    /// What a kind does when it is activated, over and above reporting a click.
    /// </summary>
    /// <remarks>
    /// A tab strip and an open dropdown are deliberately absent: the pointer already
    /// chose which tab and which option on the way down, and re-running that here would
    /// undo the choice using the release position.
    /// </remarks>
    private static void Activate(UiNode node, ActivationSource source)
    {
        switch (node.Kind)
        {
            case UiKind.Toggle:
                node.Checked = !node.Checked;
                break;

            case UiKind.Dropdown when !node.Expanded:
                node.Expanded = node.Options.Count > 0;
                break;

            case UiKind.TabStrip when source == ActivationSource.Directional:
                // Confirm on a tab strip is a no-op: left and right already moved it.
                break;
        }
    }

    private static void SetValue(UiNode node, float value)
    {
        float min = node.MinValue, max = node.MaxValue;
        if (max < min) (min, max) = (max, min);

        if (node.Step > 0f) value = min + MathF.Round((value - min) / node.Step) * node.Step;
        node.Value = Math.Clamp(value, min, max);
    }

    // -------------------------------------------------------------------------
    // Tree bookkeeping
    // -------------------------------------------------------------------------

    private UiNode? ExpandedDropdown()
    {
        if (_canvas.Root is { Kind: UiKind.Dropdown, Expanded: true }) return _canvas.Root;
        foreach (UiNode node in _canvas.Root.Descendants())
            if (node is { Kind: UiKind.Dropdown, Expanded: true, Visible: true }) return node;
        return null;
    }

    /// <summary>
    /// Clears the states that last exactly one frame, everywhere.
    /// </summary>
    /// <remarks>
    /// <see cref="UiNode.Clicked"/> in particular: a caller that reads it once a frame must
    /// never see the same click twice, and a node that stopped being hit must stop
    /// reporting a hover it can no longer have.
    /// </remarks>
    private static void ClearTransient(UiNode node)
    {
        node.Hovered = false;
        node.Clicked = false;
        foreach (UiNode child in node.Children) ClearTransient(child);
    }

    private void SyncFocusFlags(UiNode node)
    {
        node.Focused = ReferenceEquals(node, Focus.Focused);
        foreach (UiNode child in node.Children) SyncFocusFlags(child);
    }
}

File: html5/src/ui/UiInput.js
// -----------------------------------------------------------------------------
// UiInput — turns a frame of input into hover, press, click, focus and control
// behaviour on one UiCanvas.
//
// The mirror of SexyBiscuit.Engine/UI/UiInput.cs.
//
// The seven kinds after Image have no painter of their own and had no behaviour of
// their own either — a Slider was a tag that made a node focusable and nothing more.
// This is where dragging, toggling, opening and scrolling actually happen, so that a
// document describing a settings screen becomes a settings screen rather than a
// picture of one.
//
// A frame is a plain object rather than a reference to the InputManager on purpose:
// the router is then a pure function of that object and the tree, so a fixture can
// drive it with no canvas, no device and no gamepad, which is the only way the two
// engines' interaction behaviour can be pinned by one shared file.
// -----------------------------------------------------------------------------

import { UiKind, ScrollMode } from './UiEnums.js';
import { contains as rectContains } from './UiLayout.js';
import { UiFocus } from './UiFocus.js';
import { NavDirection } from './UiNavigation.js';
import { tabRect, dropdownListRect } from './UiPainter.js';

/** The fields a frame may carry, and what they mean when they are missing. */
export const EMPTY_FRAME = Object.freeze({
    deltaTime: 0,
    pointer: { x: -1, y: -1 },
    pointerDelta: { x: 0, y: 0 },
    pointerDown: false,
    pointerIsTouch: false,
    wheel: 0,
    navAxis: { x: 0, y: 0 },
    navUp: false,
    navDown: false,
    navLeft: false,
    navRight: false,
    confirm: false,
    cancel: false,
    typed: '',
    backspace: false,
});

export class UiInput {
    constructor(canvas) {
        this.canvas = canvas;
        this.focus = new UiFocus(canvas.root);

        /** The node under the pointer this frame, or null. */
        this.hovered = null;

        /** How far one wheel notch scrolls, in canvas units. */
        this.wheelStep = 48;

        /** The node the current press started on, which is the only one that can click. */
        this._pressed = null;

        /** The slider the pointer is dragging, if any. */
        this._dragging = null;

        this._wasDown = false;
    }

    // -------------------------------------------------------------------------
    // The frame
    // -------------------------------------------------------------------------

    /**
     * Resolves one frame. Call after the canvas has been laid out, so every
     * rectangle the hit test reads is this frame's.
     */
    update(input = {}) {
        const frame = { ...EMPTY_FRAME, ...input };

        clearTransient(this.canvas.root);
        this.focus.modes.tick(frame.deltaTime);

        if (!this.canvas.interactive) {
            this._wasDown = false;
            this._pressed = null;
            this._dragging = null;
            this.hovered = null;
            return;
        }

        this._noteDevices(frame);

        // Before anything reads it. Focus resolved afterwards would mean the first frame
        // a menu is open swallows its own input: a key press with nothing focused yet has
        // nowhere to go, and the player has to press twice.
        this.focus.repair();

        const hit = this._pick(frame.pointer);
        this.hovered = this.focus.modes.showHover ? hit : null;
        if (this.hovered) this.hovered.hovered = true;

        this._pointer(frame, hit);
        this._wheel(frame, hit);
        this._directional(frame);
        this._typing(frame);

        syncFocusFlags(this.canvas.root, this.focus.focused);

        this._wasDown = frame.pointerDown;
    }

    // -------------------------------------------------------------------------
    // Devices
    // -------------------------------------------------------------------------

    _noteDevices(frame) {
        if (frame.pointerIsTouch && frame.pointerDown) this.focus.modes.noteTouch();
        else if (frame.pointerDelta.x !== 0 || frame.pointerDelta.y !== 0) {
            this.focus.modes.noteMouseMotion(frame.pointerDelta);
        }

        if (frame.pointerDown && !this._wasDown && !frame.pointerIsTouch) this.focus.modes.noteMouseButton();
        if (frame.wheel !== 0) this.focus.modes.noteMouseButton();

        if (frame.navAxis.x !== 0 || frame.navAxis.y !== 0) {
            this.focus.modes.noteNavigationAxis(Math.max(Math.abs(frame.navAxis.x), Math.abs(frame.navAxis.y)));
        }

        if (frame.navUp || frame.navDown || frame.navLeft || frame.navRight || frame.confirm) {
            this.focus.modes.noteNavigationButton();
        }
    }

    // -------------------------------------------------------------------------
    // Pointer
    // -------------------------------------------------------------------------

    /**
     * The node under a device-pixel point, an open dropdown list winning over the
     * tree. A list is painted outside its control's rectangle and outside every
     * clip, so the ordinary hit test cannot see it; without this a player can see
     * the options and clicks straight through them onto whatever is behind.
     */
    _pick(devicePoint) {
        const point = this.canvas.screenToCanvas(devicePoint);
        const open = this._expandedDropdown();

        if (open && rectContains(dropdownListRect(open), point)) return open;
        return this.canvas.hitTest(devicePoint);
    }

    _pointer(frame, hit) {
        const pressedNow = frame.pointerDown && !this._wasDown;
        const releasedNow = !frame.pointerDown && this._wasDown;

        if (pressedNow) this._beginPress(frame, hit);
        else if (frame.pointerDown) this._continuePress(frame);
        else if (releasedNow) this._endPress(frame, hit);

        if (this._pressed) this._pressed.pressed = frame.pointerDown;
    }

    _beginPress(frame, hit) {
        const point = this.canvas.screenToCanvas(frame.pointer);

        // A press that lands anywhere but the open list closes it, including on the
        // control itself — which is what makes a second click on a dropdown shut it.
        const open = this._expandedDropdown();
        if (open && open !== hit) open.expanded = false;

        // A disabled node still blocks the pointer — it is not a hole — but it must not
        // press, click or take focus, so the press lands on nothing at all.
        this._pressed = usable(hit) ? hit : null;
        this._dragging = null;
        if (!this._pressed) return;

        if (UiFocus.isFocusable(hit)) this.focus.focus(hit);

        if (hit.kind === UiKind.Slider) {
            this._dragging = hit;
            setSliderFromPointer(hit, point);
        } else if (hit.kind === UiKind.TabStrip) {
            selectTabAt(hit, point);
        } else if (hit.kind === UiKind.Dropdown && hit.expanded) {
            selectOptionAt(hit, point);
        }
    }

    _continuePress(frame) {
        if (this._dragging && this._dragging.kind === UiKind.Slider) {
            setSliderFromPointer(this._dragging, this.canvas.screenToCanvas(frame.pointer));
        }
    }

    _endPress(frame, hit) {
        const pressed = this._pressed;
        this._pressed = null;
        this._dragging = null;

        if (!pressed) return;
        pressed.pressed = false;

        // The release has to land on the same node the press started on.
        if (pressed !== hit) return;

        pressed.clicked = true;
        activate(pressed, 'pointer');
    }

    // -------------------------------------------------------------------------
    // Wheel
    // -------------------------------------------------------------------------

    _wheel(frame, hit) {
        if (frame.wheel === 0 || !hit) return;

        const view = nearestScrollable(hit);
        if (!view) return;

        const max = scrollRange(view);
        const vertical = (view.scroll === ScrollMode.Vertical || view.scroll === ScrollMode.Both) && max.y > 0;

        const offset = { ...view.scrollOffset };
        if (vertical) offset.y = clamp(offset.y - frame.wheel * this.wheelStep, 0, max.y);
        else offset.x = clamp(offset.x - frame.wheel * this.wheelStep, 0, max.x);

        view.scrollOffset = offset;
    }

    // -------------------------------------------------------------------------
    // Directional
    // -------------------------------------------------------------------------

    _directional(frame) {
        if (frame.cancel) {
            const open = this._expandedDropdown();
            if (open) { open.expanded = false; return; }
        }

        const focused = this.focus.focused;

        // An open list swallows up and down: they move through the options, and only
        // once it is closed do they move between controls again.
        if (focused && focused.kind === UiKind.Dropdown && focused.expanded) {
            if (frame.navUp) stepSelection(focused, -1);
            if (frame.navDown) stepSelection(focused, +1);
            if (frame.confirm) { focused.expanded = false; focused.clicked = true; }
            return;
        }

        // A focused slider takes left and right as its own, because navigating away
        // from a volume slider by pressing right is not what anybody means by it.
        if (focused && focused.kind === UiKind.Slider && (frame.navLeft || frame.navRight)) {
            const step = focused.step > 0 ? focused.step : (focused.maxValue - focused.minValue) / 20;
            setValue(focused, focused.value + (frame.navRight ? step : -step));
            return;
        }

        // So does a tab strip, which is a row of choices however it is laid out.
        if (focused && focused.kind === UiKind.TabStrip && (frame.navLeft || frame.navRight)) {
            stepSelection(focused, frame.navRight ? +1 : -1);
            return;
        }

        if (frame.navUp) this.focus.navigate(NavDirection.Up);
        if (frame.navDown) this.focus.navigate(NavDirection.Down);
        if (frame.navLeft) this.focus.navigate(NavDirection.Left);
        if (frame.navRight) this.focus.navigate(NavDirection.Right);

        if (frame.confirm && this.focus.focused) {
            this.focus.focused.clicked = true;
            activate(this.focus.focused, 'directional');
        }
    }

    // -------------------------------------------------------------------------
    // Typing
    // -------------------------------------------------------------------------

    _typing(frame) {
        const field = this.focus.focused;
        if (!field || field.kind !== UiKind.TextField) return;

        if (frame.backspace && field.text.length > 0) field.text = field.text.slice(0, -1);
        if (frame.typed) field.text += frame.typed;
    }

    // -------------------------------------------------------------------------
    // Tree bookkeeping
    // -------------------------------------------------------------------------

    _expandedDropdown() {
        if (this.canvas.root.kind === UiKind.Dropdown && this.canvas.root.expanded) return this.canvas.root;
        for (const node of this.canvas.root.descendants()) {
            if (node.kind === UiKind.Dropdown && node.expanded && node.visible) return node;
        }
        return null;
    }
}

// -----------------------------------------------------------------------------
// Behaviour
// -----------------------------------------------------------------------------

/**
 * What a kind does when it is activated, over and above reporting a click.
 *
 * A tab strip and an open dropdown are deliberately absent: the pointer already
 * chose which tab and which option on the way down, and re-running that here would
 * undo the choice using the release position.
 */
function activate(node, source) {
    if (node.kind === UiKind.Toggle) {
        node.checked = !node.checked;
    } else if (node.kind === UiKind.Dropdown && !node.expanded) {
        node.expanded = node.options.length > 0;
    } else if (node.kind === UiKind.TabStrip && source === 'directional') {
        // Confirm on a tab strip is a no-op: left and right already moved it.
    }
}

function setSliderFromPointer(slider, point) {
    if (slider.rect.width <= 0) return;
    const f = clamp((point.x - slider.rect.x) / slider.rect.width, 0, 1);
    setValue(slider, slider.minValue + (slider.maxValue - slider.minValue) * f);
}

function setValue(node, value) {
    let min = node.minValue;
    let max = node.maxValue;
    if (max < min) [min, max] = [max, min];

    let v = value;
    if (node.step > 0) v = min + Math.round((v - min) / node.step) * node.step;
    node.value = clamp(v, min, max);
}

/** Whether a node and every ancestor of it is visible and interactive. */
function usable(node) {
    if (!node) return false;
    for (let n = node; n; n = n.parent) {
        if (!n.visible || !n.interactive) return false;
    }
    return true;
}

function selectTabAt(strip, point) {
    for (let i = 0; i < strip.options.length; i++) {
        if (rectContains(tabRect(strip, i), point)) { strip.selectedIndex = i; return; }
    }
}

function selectOptionAt(dropdown, point) {
    const list = dropdownListRect(dropdown);
    if (dropdown.rect.height <= 0 || !rectContains(list, point)) return;

    const index = Math.floor((point.y - list.y) / dropdown.rect.height);
    if (index < 0 || index >= dropdown.options.length) return;

    dropdown.selectedIndex = index;
    dropdown.expanded = false;
    dropdown.clicked = true;
}

function stepSelection(node, by) {
    const count = node.options.length;
    if (count === 0) return;
    node.selectedIndex = clamp(node.selectedIndex + by, 0, count - 1);
}

/** The first ancestor that scrolls and has something to scroll over. */
function nearestScrollable(from) {
    for (let n = from; n; n = n.parent) {
        if (n.scroll === ScrollMode.None) continue;
        const range = scrollRange(n);
        if (range.x > 0 || range.y > 0) return n;
    }
    return null;
}

function scrollRange(view) {
    return {
        x: Math.max(0, view.contentSize.x - view.contentRect.width),
        y: Math.max(0, view.contentSize.y - view.contentRect.height),
    };
}

/**
 * Clears the states that last exactly one frame, everywhere. `clicked` in
 * particular: a caller that reads it once a frame must never see the same click
 * twice, and a node that stopped being hit must stop reporting a hover.
 */
function clearTransient(node) {
    node.hovered = false;
    node.clicked = false;
    for (const child of node.children) clearTransient(child);
}

function syncFocusFlags(node, focused) {
    node.focused = node === focused;
    for (const child of node.children) syncFocusFlags(child, focused);
}

function clamp(v, min, max) { return Math.min(max, Math.max(min, v)); }
